File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetVatTu.java
package vn.ditagis.com.tanhoa.qlsc.services;

import android.content.Context;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;

import vn.ditagis.com.tanhoa.qlsc.entities.Constant;
import vn.ditagis.com.tanhoa.qlsc.entities.DLayerInfo;
import vn.ditagis.com.tanhoa.qlsc.entities.VatTu;
import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB;


public class GetVatTu extends AsyncTask<Void, Boolean, Void> {
    private Context mContext;
    private AsyncResponse mDelegate;

    public interface AsyncResponse {
        void processFinish();
    }


    public GetVatTu(Context context, AsyncResponse delegate) {
        this.mContext = context;
        this.mDelegate = delegate;
    }


    private void getVatTuFromService() {

        String layerInfoVatTu = Constant.IDLayer.ID_VAT_TU_TABLE;

        for (DLayerInfo dLayerInfo : ListObjectDB.getInstance().getLstFeatureLayerDTG()) {
            if (dLayerInfo.getId().equals(layerInfoVatTu)) {
                final QueryParameters queryParameters = new QueryParameters();
                queryParameters.setWhereClause("1=1");
                String url = dLayerInfo.getUrl();
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.getUrl();
                final ServiceFeatureTable serviceFeatureTable = new ServiceFeatureTable(url);
//
                final ListenableFuture<FeatureQueryResult> feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
                feature.addDoneListener(() -> {
                    final List<VatTu> vatTuList = new ArrayList<>();
                    try {
                        FeatureQueryResult result = feature.get();
                        Iterator<Feature> iterator = result.iterator();
                        Feature item;
                        while (iterator.hasNext()) {
                            item = iterator.next();
                            String maVatTu = (String) item.getAttributes().get(Constant.FIELD_VATTU.MA_VAT_TU);
                            String tenVatTu = (String) item.getAttributes().get(Constant.FIELD_VATTU.TEN_VAT_TU);
                            String donViTinh = (String) item.getAttributes().get(Constant.FIELD_VATTU.DON_VI_TINH);
                            VatTu vatTu = new VatTu(maVatTu, tenVatTu, donViTinh);
                            vatTuList.add(vatTu);
                        }
                        ListObjectDB.getInstance().setVatTus(vatTuList);
                        publishProgress(true);

                    } catch (InterruptedException | ExecutionException e) {
                        e.printStackTrace();
                        publishProgress();
                    }


                });

                break;
            }
        }

    }


    @Override
    protected Void doInBackground(Void... voids) {
        getVatTuFromService();
        return null;
    }

    @Override
    protected void onProgressUpdate(Boolean... values) {
        super.onProgressUpdate(values);
        if(values!= null && values.length > 0 && values[0])
            mDelegate.processFinish();
    }
}

File: app/src/main/java/vn/ditagis/com/tanhoa/qlsc/services/GetVatTu.kt
package vn.ditagis.com.tanhoa.qlsc.services

import android.content.Context
import android.os.AsyncTask
import android.widget.Toast

import com.esri.arcgisruntime.concurrent.ListenableFuture
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.FeatureQueryResult
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable

import java.util.ArrayList
import java.util.concurrent.ExecutionException

import vn.ditagis.com.tanhoa.qlsc.entities.Constant
import vn.ditagis.com.tanhoa.qlsc.entities.DApplication
import vn.ditagis.com.tanhoa.qlsc.entities.DLayerInfo
import vn.ditagis.com.tanhoa.qlsc.entities.VatTu
import vn.ditagis.com.tanhoa.qlsc.entities.entitiesDB.ListObjectDB
import java.lang.Exception


class GetVatTu(private val mApplication: DApplication, private val mDelegate: AsyncResponse) : AsyncTask<Void, Boolean, Void>() {

    interface AsyncResponse {
        fun processFinish()
    }


    private fun getVatTuFromService() {

        val layerInfoVatTu = Constant.IDLayer.ID_VAT_TU_TABLE

        for (dLayerInfo in mApplication.lstFeatureLayerDTG!!) {
            if (dLayerInfo.id == layerInfoVatTu) {
                val queryParameters = QueryParameters()
                queryParameters.whereClause = "1=1"
                var url = dLayerInfo.url
                if (!url.startsWith("http"))
                    url = "http:" + dLayerInfo.url
                val serviceFeatureTable = ServiceFeatureTable(url)
                //
                val feature = serviceFeatureTable.queryFeaturesAsync(queryParameters, ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
                feature.addDoneListener {
                    val vatTuList = ArrayList<VatTu>()
                    try {
                        val result = feature.get()
                        val iterator = result.iterator()
                        var item: Feature
                        while (iterator.hasNext()) {
                            item = iterator.next()
                            val maVatTu = item.attributes[Constant.FIELD_VATTU.MA_VAT_TU] as String
                            val tenVatTu = item.attributes[Constant.FIELD_VATTU.TEN_VAT_TU] as String
                            val donViTinh = item.attributes[Constant.FIELD_VATTU.DON_VI_TINH] as String
                            val vatTu = VatTu(maVatTu, tenVatTu, donViTinh)
                            vatTuList.add(vatTu)
                        }
                        ListObjectDB.instance.vatTus = vatTuList
                        publishProgress(true)

                    } catch (e: InterruptedException) {
                        publishProgress()
                    } catch (e: ExecutionException) {
                        publishProgress()
                    }


                }

                break
            }
        }

    }


    override fun doInBackground(vararg voids: Void): Void? {
        try {
            getVatTuFromService()
        } catch (ex: Exception) {
//            Toast.makeText(mApplication, "Có lỗi xảy ra khi lấy vật tư", Toast.LENGTH_SHORT).show()
        }
        return null
    }

    override fun onProgressUpdate(vararg values: Boolean?) {
        super.onProgressUpdate(*values)
        if (values.isNotEmpty() && values[0]!!)
            mDelegate.processFinish()
    }
}
